File: Coocoo3DGraphics/Uploader.cpp
#include "pch.h"
#include "Uploader.h"
#include "DirectXHelper.h"
using namespace Coocoo3DGraphics;
using namespace Microsoft::WRL;

void Uploader::Texture2DRaw(const Platform::Array<byte>^ rawData, Format format, int width, int height)
{
	Texture2DRaw(rawData, format, width, height, 1);
}

void Uploader::Texture2DRaw(const Platform::Array<byte>^ rawData, Format format, int width, int height, int mipLevel)
{
	this->width = width;
	this->height = height;
	this->m_format = (DXGI_FORMAT)format;
	this->mipLevels = mipLevel;
	this->m_data = std::vector<byte>();
	this->m_data.resize(rawData->Length);
	memcpy(m_data.data(), rawData->begin(), rawData->Length);
}

void Uploader::Texture2DPure(int width, int height, Windows::Foundation::Numerics::float4 color)
{
	this->width = width;
	this->height = height;
	this->m_format = DXGI_FORMAT_R32G32B32A32_FLOAT;
	this->mipLevels = 1;
	int count = width * height;
	this->m_data.resize(count * 16);

	void* p = m_data.data();
	float* p1 = (float*)p;
	for (int i = 0; i < count; i++) {
		*p1 = color.x;
		*(p1 + 1) = color.y;
		*(p1 + 2) = color.z;
		*(p1 + 3) = color.w;
		p1 += 4;
	}
}

void Uploader::TextureCubeRaw(const Platform::Array<byte>^ rawData, Format format, int width, int height, int mipLevel)
{
	this->width = width;
	this->height = height;
	this->m_format = (DXGI_FORMAT)format;
	this->mipLevels = mipLevel;
	this->m_data = std::vector<byte>();
	this->m_data.resize(rawData->Length);
	memcpy(this->m_data.data(), rawData->begin(), rawData->Length);
}

void Uploader::TextureCubePure(int width, int height, const Platform::Array<Windows::Foundation::Numerics::float4>^ color)
{
	this->width = width;
	this->height = height;
	this->m_format = DXGI_FORMAT_R32G32B32A32_FLOAT;
	this->mipLevels = 1;
	int count = width * height;
	if (count < 256)throw ref new Platform::NotImplementedException("Texture too small");
	this->m_data.resize(count * 16 * 6);

	void* p = this->m_data.data();
	float* p1 = (float*)p;
	for (int c = 0; c < 6; c++)
	{
		for (int i = 0; i < count; i++) {
			*p1 = color[c].x;
			*(p1 + 1) = color[c].y;
			*(p1 + 2) = color[c].z;
			*(p1 + 3) = color[c].w;
			p1 += 4;
		}
	}
}

File: Coocoo3DGraphics1/Uploader.cs
using System;
using System.Collections.Generic;
using System.Numerics;
using System.Runtime.InteropServices;
using System.Text;
using Vortice.DXGI;
using static Coocoo3DGraphics.DXHelper;

namespace Coocoo3DGraphics
{
    public class Uploader
    {
        public int m_width;
        public int m_height;
        public int m_mipLevels;
        public Format m_format;

        public byte[] m_data;
        public void Texture2DRaw(byte[] rawData, Format format, int width, int height)
        {
            m_width = width;
            m_height = height;
            m_format = format;
            m_mipLevels = 1;
            m_data = new byte[rawData.Length];
            Array.Copy(rawData, m_data, rawData.Length);
        }

        public void Texture2DRaw(byte[] rawData, Format format, int width, int height, int mipLevel)
        {
            m_width = width;
            m_height = height;
            m_format = format;
            m_mipLevels = mipLevel;
            m_data = new byte[rawData.Length];
            Array.Copy(rawData, m_data, rawData.Length);
        }

        public void Texture2DPure(int width, int height, Vector4 color)
        {
            m_width = width;
            m_height = height;
            m_format = Format.R32G32B32A32_Float;
            m_mipLevels = 1;
            int count = width * height;
            m_data = new byte[count * 16];
            var d1 = MemoryMarshal.Cast<byte, Vector4>(m_data);
            for (int i = 0; i < d1.Length; i++)
            {
                d1[i] = color;
            }
        }

        public void TextureCubeRaw(byte[] rawData, Format format, int width, int height, int mipLevel)
        {

            m_width = width;
            m_height = height;
            m_format = format;
            m_mipLevels = mipLevel;
            m_data = new byte[rawData.Length];
            Array.Copy(rawData, m_data, rawData.Length);
        }

        public void TextureCubePure(int width, int height, Vector4[] color)
        {
            m_width = width;
            m_height = height;
            m_format = Format.R32G32B32A32_Float;
            m_mipLevels = 1;
            int count = width * height;
            if (count < 256) throw new NotImplementedException("Texture too small");
            m_data = new byte[count * 16 * 6];
            var d1 = MemoryMarshal.Cast<byte, Vector4>(m_data);
            for (int j = 0; j < 6; j++)
                for (int i = 0; i < count; i++)
                {
                    d1[i + j * count] = color[j];
                }
        }
    }
}
